WIP

jantimar opened the pull request:
Add CLLocationManagerProtocol
Remove unused WMFMockLocationManager
Remove static and class func from WMFLocationManager
Add tests for WMFLocationManager

jantimar committed:
Remove unused WMFMockLocationManager

jantimar committed:
Change WMFLocationManager class functions to instance functions

jantimar committed:
Move the `LocationManager.isAuthorized` check from the cell

VojtaStavik commented on Wikipedia.xcodeproj/project.pbxproj:
```diff
@@ -1717,7 +1717,6 @@
 		BAC6EEC91F1E519B00228AD0 /* AppearanceSettingsViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = BAFCE8411F1D7FD30077D5E9 /* AppearanceSettingsViewController.swift */; };
```
I think this commit should be split in two - `Remove unused WMFMockLocationManager` and `Remove unused code`.

VojtaStavik commented on Wikipedia/Code/ArticleLocationAuthorizationCollectionViewCell.swift:
```diff
@@ -41,9 +41,6 @@ class ArticleLocationAuthorizationCollectionViewCell: ArticleLocationExploreColl
     }
 
     public func updateForLocationEnabled() {
-        guard WMFLocationManager.isAuthorized() else {
```
This functionality is just removed? Why?

jantimar replied:
This function is call from 2 places

- ExploreCardViewController - locationManager -> https://github.com/industrialbinaries/wikipedia-ios/pull/2#discussion_r382517890

- ExploreCardViewController - configureLocationCell -> where this check is

IMHO: I think this check is better when it is in ViewController as in Cell

VojtaStavik commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -38,11 +41,11 @@ NS_ASSUME_NONNULL_BEGIN
-
-+ (BOOL)isAuthorizationNotDetermined;
-+ (BOOL)isAuthorizationDenied;
-+ (BOOL)isAuthorizationRestricted;
```
This change is here because you need to use the location manager ivar?

jantimar replied:
To remove static properties from LocaionManager when it is not required

VojtaStavik commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -1,4 +1,5 @@
 @import CoreLocation;
+@protocol WMFCLLocationManagerProtocol;
```
What is this protocol?

VojtaStavik commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -67,7 +65,7 @@ + (instancetype)coarseLocationManager {
     self = [super init];
     if (self) {
-        self.currentAuthorizationStatus = [CLLocationManager authorizationStatus];
+        self.currentAuthorizationStatus = [[locationManager class] authorizationStatus];
```
👍

jantimar replied:
Separate commit

VojtaStavik commented on Wikipedia/Code/ExploreCardViewController.swift:
```diff
@@ -663,7 +663,7 @@ extension ExploreCardViewController: WMFLocationManagerDelegate {
         UserDefaults.wmf.wmf_setLocationAuthorized(enabled)
         for cell in collectionView.visibleCells {
-            guard let cell = cell as? ArticleLocationAuthorizationCollectionViewCell else {
+            guard let cell = cell as? ArticleLocationAuthorizationCollectionViewCell, locationManager.isAuthorized() else {
```
There's a new functionality here? Why? The commit is named `Remove static and class func from WMFLocationManager`

VojtaStavik commented on Wikipedia/Code/NSUserDefaults+WMFExtensions.swift:
```diff
@@ -193,7 +193,6 @@ let WMFSearchLanguageKey = "WMFSearchLanguageKey"
         return self.bool(forKey: WMFLocationAuthorizedKey)
     }
     
-    
```
This shouldn't be in this commit

VojtaStavik commented on WikipediaUnitTests/CLHeadingMock.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+/// Mock `CLHeading` with enable update heading accuracy
+final class CLHeadingMock: CLHeading {
```
Just an idea - what if we use this patter for mocks and stubs? `CLHeading_Mock`

jantimar replied:
Resolved with `MockCLHeading`

VojtaStavik commented on WikipediaUnitTests/CLHeadingMock.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+/// Mock `CLHeading` with enable update heading accuracy
```
This sentence doesn't make sense 🤷‍♂ Maybe something like:
`A CLHeading subclass allowing modification of the headingAccuracy  value.`

VojtaStavik commented on WikipediaUnitTests/CLLocationManagerMock.swift:
```diff
@@ -0,0 +1,55 @@
+import Foundation
+
+/// Mock for `CLLocationManager`
```
`A CLLocationManager subclass allowing mocking in tests.`

VojtaStavik commented on WikipediaUnitTests/CLLocationManagerMock.swift:
```diff
@@ -0,0 +1,55 @@
+/// Mock for `CLLocationManager`
+final class CLLocationManagerMock: CLLocationManager {
+
+    // MARK: - Override CLLocationManagerProtocol
```
???

VojtaStavik commented on WikipediaUnitTests/CLLocationManagerMock.swift:
```diff
@@ -0,0 +1,55 @@
+    // MARK: - Override CLLocationManagerProtocol
+
+    var _heading: CLHeading?
+    override var heading: CLHeading? {
```
```
    override var heading: CLHeading? {  _heading }
```
???

VojtaStavik commented on WikipediaUnitTests/CLLocationManagerMock.swift:
```diff
@@ -0,0 +1,55 @@
+        return _location
+    }
+
+    override func requestAlwaysAuthorization() { }
```
Why are these overrides empty?

jantimar replied:
Because in production CLLocationManager on new installation can require user interaction - permission prompt

VojtaStavik commented on WikipediaUnitTests/CLLocationManagerMock.swift:
```diff
@@ -0,0 +1,55 @@
+    var updatingLocation: Bool?
+    var updatingHeading: Bool?
+
+    static func resetMockData() {
```
It would be nice to have a comment here explaining what this method does.

jantimar replied:
`Reset CLLocationManager to the default values:
- `

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,132 @@
+
+    func testStopMonitoring() {
+        let mock = CLLocationManagerMock()
+        let locationManager = WMFLocationManager(locationManager: mock)
```
Why do you create new `mock` and `locationManager` here?

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,132 @@
+        XCTAssertEqual(locationManager.isAuthorizationRestricted(), true)
+    }
+
+    // MARK: - WMFLocationManagerDelegate tests
```
These tests don't make sense. Let's talk IRL.

VojtaStavik reviewed (commented):
I didn't manage to review the last commit yet. Let's talk first.

jantimar committed:
Remove unused property `significantLocationUpdatesOnly`
This property was never set.

jantimar commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -7,8 +7,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface WMFLocationManager : NSObject
 
-@property (nonatomic, strong, readonly) CLLocationManager *locationManager;
```
Separate to 3 commits

jantimar commented on Wikipedia/Code/ArticleLocationAuthorizationCollectionViewCell.swift:
```diff
@@ -41,9 +41,6 @@ class ArticleLocationAuthorizationCollectionViewCell: ArticleLocationExploreColl
     }
 
     public func updateForLocationEnabled() {
```
Move `isAuthorized` check from the cell to the view controller

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -97,26 +97,30 @@ - (NSString *)description {
 #pragma mark - Permissions
 
-+ (BOOL)isAuthorized {
++ (BOOL)isAuthorized DEPRECATED_MSG_ATTRIBUTE("Use `- (BOOL)isAuthorized`") {
```
Delete!

jantimar commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -19,6 +19,8 @@ NS_ASSUME_NONNULL_BEGIN
 
 + (instancetype)coarseLocationManager;
```
Make the initializer with the location manager public
To enable dependency injection

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -118,7 +118,6 @@ - (BOOL)isAuthorizationRestricted {
 - (BOOL)requestAuthorizationIfNeeded {
     CLAuthorizationStatus status = [[self.locationManager class] authorizationStatus];
     if (status == kCLAuthorizationStatusNotDetermined) {
-        NSParameterAssert([[self.locationManager class] locationServicesEnabled]);
```
😱

jantimar commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,129 @@
+
+        let location = CLLocation(latitude: 10, longitude: 20)
+        mock._location = location
+        mock.delegate?.locationManager?(mock, didUpdateLocations: [])
```
`mock.simulateLocationUpdate(location)`

jantimar commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    func testChangeEnabledState() {
+        type(of: mock)._authorizationStatus = .authorizedAlways
```
`mock.simulate ...`

jantimar commented on WMF Framework/LocationManagerProtocol.swift:
```diff
@@ -0,0 +1,60 @@
+    case authorized, notDeterminded, denied, restricted, unknown
+}
+
+@objc public protocol LocationManagerProtocol {
```
Add LocationManagerProtocol
This protocol mimics the WMFLocationManager header file. It provides an Objective-C compatible API to the Swift version of LocationManager coming in the next commit.

jantimar commented on Wikipedia/Code/ArticleLocationCollectionViewController.swift:
```diff
@@ -155,7 +155,21 @@ extension ArticleLocationCollectionViewController {
 // MARK: - WMFLocationManagerDelegate
-extension ArticleLocationCollectionViewController: WMFLocationManagerDelegate {
+extension ArticleLocationCollectionViewController: LocationManagerDelegate {
+    func didUpdate(location: CLLocation, from locationManager: LocationManagerProtocol) {
```
follow cocoa delegate method patter

jantimar committed:
Remove unused method `restartLocationMonitoring`

jantimar committed:
Get CLLocationManager type dynamically

jantimar committed:
Add ObjC support for LocationManager

jantimar committed:
Remove WMFLocationManager

jantimar committed:
Replace WMFLocationManager with LocationManager

jantimar commented on Wikipedia/Code/WMFAppViewController.m:
```diff
@@ -876,7 +876,8 @@ - (void)finishResumingApp {
 
-    BOOL locationAuthorized = [WMFLocationManager isAuthorized];
+    CLAuthorizationStatus status = [CLLocationManager authorizationStatus];
+    BOOL locationAuthorized = status == kCLAuthorizationStatusAuthorizedAlways || status == kCLAuthorizationStatusAuthorizedWhenInUse;
```
`[LocationManagerFactory coarseLocationManager].state == LocationManagerStateAuthorized` ??

jantimar committed:
Move reverse geocoding out of LocationManager

jantimar committed:
Add WMFLocationManager testing category

This makes the initialazer allowing dependency injection accessible from the unit test target.

VojtaStavik commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -40,11 +40,10 @@ NS_ASSUME_NONNULL_BEGIN
-
-+ (BOOL)isAuthorizationNotDetermined;
-+ (BOOL)isAuthorizationDenied;
-+ (BOOL)isAuthorizationRestricted;
```
Tohle je jen zmena formatovani, nema to co delat v tomhle commitu,

VojtaStavik commented on WikipediaUnitTests/MockCLLocationManager.swift:
```diff
@@ -0,0 +1,82 @@
+        updatingHeading = false
+    }
+
+    // Override method which can required user interaction - f.e. permission
```
method**s**

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -0,0 +1,241 @@
+import UIKit
+import CoreLocation
+
+final public class LocationManager: NSObject, LocationManagerProtocol {
```
Proc je tohle `NSObject`?

jantimar replied:
-> `CLLocationManagerDelegate`

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    /// Call when device update orientation heading,
+    @objc private func updateDeviceHeading() {
```
Proc je tophle `@objc`?

jantimar replied:
-> 
```
NotificationCenter.default.addObserver(
            self,
            selector: #selector(updateDeviceHeading),
            name: UIDevice.orientationDidChangeNotification,
            object: nil
        )
```

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,144 @@
+
+    func testFineLocationManager() {
+        let locationManager = WMFLocationManager.fine()
+        XCTAssertEqual(locationManager.locationManager.distanceFilter, 1)
```
`locationManager.locationManager` 😱

jantimar replied:
WMF code ¯\_(ツ)_/¯

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,144 @@
+}
+
+/// Test implementation of `WMFLocationManagerDelegate`
+private final class LocationManagerTestsDelegate: NSObject, WMFLocationManagerDelegate {
```
`TestLocationManagerDelegate` ?

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,144 @@
+        locationManager.startMonitoringLocation()
+        XCTAssertEqual(locationManager.isUpdating, true)
+        XCTAssertEqual(mock.updatingHeading, true)
+        XCTAssertEqual(mock.updatingHeading, true)
```
Same line as above

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,144 @@
+        locationManager.stopMonitoringLocation()
+        XCTAssertEqual(locationManager.isUpdating, false)
+        XCTAssertEqual(mock.updatingHeading, false)
+        XCTAssertEqual(mock.updatingHeading, false)
```
duplicated line

jantimar replied:
🙀

VojtaStavik commented on WikipediaUnitTests/MockCLLocationManager.swift:
```diff
@@ -0,0 +1,82 @@
+
+    // MARK: - Test properties
+
+    var updatingLocation: Bool?
```
`isUpdatingLocation`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,144 @@
+        super.setUp()
+
+        MockCLLocationManager.resetMockData()
+        mock = MockCLLocationManager()
```
`mock` is not enough. Maybe `mockCLLocationManager` is better. There are too many location managers and it's not clear what is what.

VojtaStavik commented on WikipediaUnitTests/MockCLLocationManager.swift:
```diff
@@ -0,0 +1,82 @@
+
+    /// Update authorization status to new value, this value will be rewrite in all `CLLocationManagerMock`
+    /// - Parameter authorizationStatus: New authorization status
+    func update(authorizationStatus: CLAuthorizationStatus) {
```
Why is this `update` when the others are `simulate`?

jantimar replied:
Other simulate the behavior and call `delegate`
this one just update value

jantimar committed:
Add LocationManager skeleton and update LocationManagerTests

jantimar committed:
Implement LocationManager functionality

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -0,0 +1,244 @@
+    }
+    // TODO: Temporary helper for support ObjC codebase, leter can call directly `autorizationStatus.authorized`
+    /// Return `true` if user is aurthorized or authorized always
+    public var isAuthorized: Bool {
```
Muzeme dat tohle do toho souboru kde definujeme protocol jako extension?

VojtaStavik replied:
Nebo aspon do samostatny extension tady?

jantimar replied:
Nie, používa sa v ObjC

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -1,6 +1,10 @@
 import UIKit
 import CoreLocation
 
+public enum LocationManagerError: Error {
```
Proc tenhle enum nenim uvnitr `LocationManager`?

jantimar replied:
Enum odstraneny keďže bol sučasťou presunuteho geocode

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -1,6 +1,10 @@
 import CoreLocation
 
+public enum LocationManagerError: Error {
+    case missingPlacemark
```
Tady bysme meli pridat nejaky comment

jantimar replied:
Enum odstraneny keďže bol sučasťou presunuteho geocode

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -16,29 +20,225 @@ extension LocationManagerAccuracy {
+                startMonitoringLocation()
+                waitingForAuthorization = false
+            }
+        @unknown default: break
```
Tohle bych dal pryc.

jantimar replied:
Jedna sa o objc enum :(

VojtaStavik commented on WMF Framework/LocationManager.swift:
```diff
@@ -16,29 +20,225 @@ extension LocationManagerAccuracy {
         type: LocationManagerAccuracy = .fine
-    ) { }
+    ) {
+        locationManager.activityType = .fitness
```
Let's also add back the missing `    NSAssert([NSThread isMainThread], @"Location managers must be created on the main thread");`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -52,6 +52,15 @@ final class LocationManagerTests: XCTestCase {
 
+    // MARK: - Authorization
+
+    func testRequestAuthorization() {
```
`testStartMonitoringCalledWhenAuthorizationSuccessfull`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -84,6 +93,12 @@ final class LocationManagerTests: XCTestCase {
         XCTAssertEqual(delegate.location, location)  // Test new location on delegate
     }
 
+    func testLastKnowLocation() {
```
`testExistingLocationWithoutCallback`
// When the location has already been fetched, a new instance of CLLocationManager contains the last known location in its `location` variable, even before `startMonitoringLocation()` has been called.

jantimar committed:
Make WMFLocationManager's UIDevice dependency explicit

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        XCTAssertEqual(mockCLLocationManager.isUpdatingHeading, true)
+    }
+
+    func testStartLocationWithoutPermission() {
```
??

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+        mockCLLocationManager.update(authorizationStatus: .restricted)
+        locationManager.startMonitoringLocation()
+        XCTAssertEqual(locationManager.isUpdating, false)
```
add mockCLLocationManager checks

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+        mockCLLocationManager.update(authorizationStatus: .authorizedAlways)
+        locationManager.startMonitoringLocation()
+        XCTAssertEqual(locationManager.isUpdating, true)
```
Delete

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+    // MARK: - Authorization
+
+    func testStartMonitoringCalledWhenAuthorizationSuccessfull() {
```
testPermissionRequestedWhenAuthorizationStatusNotDetermined

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+        XCTAssertEqual(mockCLLocationManager.isRequestedForAuthorization, true)
+    }
+
```
missing `testStartMonitoringCalledWhenAuthorizationSuccessfull`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        mockCLLocationManager.update(authorizationStatus: .notDetermined)
+        locationManager.startMonitoringLocation()
+
+        XCTAssertEqual(mockCLLocationManager.isRequestedForAuthorization, true)
```
-> authorization callback called -> starts monitoring

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+    // When the location has already been fetched, a new instance of CLLocationManager contains the last known location in its location variable, even before startMonitoringLocation() has been called
+    func testExistingLocationWithoutCallback() {
+        let location = CLLocation(latitude: 10, longitude: 20)
```
`// startMonitoringLocation() not called`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+    func testExistingLocationWithoutCallback() {
+        let location = CLLocation(latitude: 10, longitude: 20)
+        mockCLLocationManager.simulateUpdate(location: location)
+        XCTAssertEqual(locationManager.location, location)
```
add `let locationManager = LocationManager(....)`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        let heading2 = MockCLHeading(headingAccuracy: 100)
+        mockCLLocationManager.simulateUpdate(heading: heading2)
+
+        // Heading and location
```
// Check the values are not updated

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        XCTAssertEqual(locationManager.location, location1)
+        XCTAssertEqual(delegate.location, location1)
+
+        // Error
```
// Check the error is not propagated

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+    func testChangeAuthorizedState() {
+        mockCLLocationManager.update(authorizationStatus: .authorizedAlways)
+        mockCLLocationManager.delegate?.locationManager?(CLLocationManager(), didChangeAuthorization: .denied)
```
`mockCLLocationManager.delegate?.locationManager?(mockCLLocationManager, didChangeAuthorization: .denied)`

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    func testChangeAuthorizedState() {
+        mockCLLocationManager.update(authorizationStatus: .authorizedAlways)
```
- probably redundant
- replace with simulate() ?

VojtaStavik commented on WikipediaUnitTests/MockUIDevice.swift:
```diff
@@ -0,0 +1,21 @@
+/// A UIDevice subclass allowing mocking in tests.
+final class MockUIDevice: UIDevice {
+
+    private var _orientation: UIDeviceOrientation = UIDevice.current.orientation
```
init(orientation: .unknown)

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        locationManager.startMonitoringLocation()
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
+        mockDevice.simulateUpdate(orientation: .unknown)
```
redundant

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+    func testUpdateDeviceHeading() {
+        locationManager.startMonitoringLocation()
+        let heading = MockCLHeading(headingAccuracy: 10)
+        mockCLLocationManager.simulateUpdate(heading: heading)
```
redundant?

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+        mockDevice.simulateUpdate(orientation: .landscapeLeft)
+        XCTAssertEqual(mockCLLocationManager.headingOrientation, .landscapeLeft)
+
+        locationManager.stopMonitoringLocation()
```
// Device orientation updates should not be propagated when the monitoring is stopped.

VojtaStavik commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,230 @@
+
+        locationManager.stopMonitoringLocation()
+        mockDevice.simulateUpdate(orientation: .portrait)
+        XCTAssertEqual(mockCLLocationManager.headingOrientation, .landscapeLeft)
```
`XCTAssertNotEqual(mockCLLocationManager.headingOrientation, .portrait)`

VojtaStavik commented on WikipediaUnitTests/MockCLLocationManager.swift:
```diff
@@ -0,0 +1,85 @@
+
+    /// Reset `CLLocationManagerMock` to the default values:
+    /// - authorizationStatus = .authorizedAlways
+    static func resetMockData() {
```
remove

jantimar committed:
Add logs to LocationManager

jantimar commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -32,8 +32,6 @@ NS_ASSUME_NONNULL_BEGIN
 /**
  *  Start monitoring location and heading updates.
```
add `This method is idempotent.`

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -14,8 +14,6 @@
 
 @interface WMFLocationManager () <CLLocationManagerDelegate>
```
Commit message:

Remove unused property `significantLocationUpdatesOnly`
This property was never set.

jantimar commented on Wikipedia/Code/ArticleLocationAuthorizationCollectionViewCell.swift:
```diff
@@ -41,9 +41,6 @@ class ArticleLocationAuthorizationCollectionViewCell: ArticleLocationExploreColl
     }
```
message: 

Make the check in `ExploreCardViewController` instead of accessing `WMFLocationManager` from the cell.

jantimar commented on Wikipedia/Code/ArticleLocationCollectionViewController.swift:
```diff
@@ -36,7 +36,7 @@ class ArticleLocationCollectionViewController: ColumnarCollectionViewController,
     override func viewWillAppear(_ animated: Bool) {
```
Change WMFLocationManager class functions to instance functions

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -16,6 +16,7 @@ @interface WMFLocationManager () <CLLocationManagerDelegate>
 
 @property (nonatomic, strong, readwrite) CLLocationManager *locationManager;
```
Move before "Add WMFLocationManager testing category"

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -16,6 +16,7 @@ @interface WMFLocationManager () <CLLocationManagerDelegate>
 
 @property (nonatomic, strong, readwrite) CLLocationManager *locationManager;
 @property (nonatomic, strong, nullable) id orientationNotificationToken;
```
message: Make WMFLocationManager's UIDevice dependency explicit

jantimar commented on Wikipedia/Code/WMFLocationManager.m:
```diff
@@ -51,8 +51,6 @@ @interface WMFLocationManager () <CLLocationManagerDelegate>
 
 @property (nonatomic, assign, readwrite) CLAuthorizationStatus currentAuthorizationStatus;
 
-- (instancetype)initWithLocationManager:(CLLocationManager *)locationManager device:(UIDevice *)device NS_DESIGNATED_INITIALIZER;
```
??? previous commit?

jantimar commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -0,0 +1,300 @@
+import XCTest
+@testable import WMF
+
```
Try to split to more commits

jantimar commented on WMF Framework/LocationManagerProtocol.swift:
```diff
@@ -0,0 +1,34 @@
+import CoreLocation
```
Move after "Add logs to LocationManager"

jantimar replied:
maybe merge with "Add ObjC LocationManagerFactory" and rename to "Add ObjC support for LocationManager"

jantimar commented on Wikipedia/Code/WMFLocationManager.h:
```diff
@@ -44,10 +44,6 @@ NS_ASSUME_NONNULL_BEGIN
 - (BOOL)isAuthorizationNotDetermined;
 - (BOOL)isAuthorizationDenied;
 - (BOOL)isAuthorizationRestricted;
-
```
Move after "Remove unused method `restartLocationMonitoring`"

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -0,0 +1,46 @@
+    public var autorizationStatus: CLAuthorizationStatus { fatalError() }
+    // TODO: Temporary helper for support ObjC codebase, leter can call directly `autorizationStatus.authorized`
+    /// Return `true` if user is aurthorized or authorized always
+    public var isAuthorized: Bool { fatalError() }
```
empty lines 😱

jantimar commented on WikipediaUnitTests/Code/LocationManagerTests.swift:
```diff
@@ -245,14 +241,14 @@ final class LocationManagerTests: XCTestCase {
         // Test stop monitoring disable
         locationManager.stopMonitoringLocation()
         XCTAssertEqual(mockDevice.beginGeneratingDeviceOrientationCount, 1)
-//        XCTAssertEqual(mockDevice.endGeneratingDeviceOrientationCount, 1)
```
missing - currently failing

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -17,30 +17,180 @@ extension LocationManagerConfiguration {
 final public class LocationManager: NSObject, LocationManagerProtocol {
 
     /// Last know location
```
Implement LocationManager functionality

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -17,30 +17,180 @@ extension LocationManagerConfiguration {
+        }
+
+        guard isAuthorized else {
+            authorizedCompletion = startMonitoringLocation
```
// Start monitoring location in case the authorization status ever changes to `authorized`

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -17,30 +17,180 @@ extension LocationManagerConfiguration {
+
+    // TODO: Temporary helper for support ObjC codebase, leter can call directly `autorizationStatus.authorized`.
+    /// Return `true` if user is aurthorized or authorized always.
+    public var isAuthorized: Bool { autorizationStatus.isAuthorized }
```
Move to "Add ObjC support ..."

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -17,30 +17,180 @@ extension LocationManagerConfiguration {
+    }
+}
+
+private extension CLAuthorizationStatus {
```
MAke public

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -17,30 +17,180 @@ extension LocationManagerConfiguration {
+
+    private func authorize(succcess: (() -> Void)? = nil) {
+        locationManager.requestWhenInUseAuthorization()
+        authorizedCompletion = succcess
```
swap lines

jantimar commented on WMF Framework/LocationManager.swift:
```diff
@@ -1,5 +1,6 @@
 import UIKit
```
Add "Don't propagate location errors when in the simulator" functionality

jantimar commented on Wikipedia/Code/ArticleLocationCollectionViewController.swift:
```diff
@@ -107,6 +107,15 @@ class ArticleLocationCollectionViewController: ColumnarCollectionViewController,
             ReadingListsFunnel.shared.logUnsaveInFeed(context: context, articleURL: articleURL, index: previewedIndexPath?.item)
```
Move somewhere close to the beginning

jantimar replied:
subject: Move non-delegate code out of the delegate extension

jantimar committed:
Move non-delegate code out of the delegate extension

jantimar committed:
Don't propagate location errors when in the simulator

jantimar committed:
Add basic LocationManager tests

jantimar committed:
Add device orientation LocationManager tests
